File: app/src/main/java/net/syncthing/lite/adapters/DevicesAdapter.java
package net.syncthing.lite.adapters;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.graphics.PorterDuff;
import android.support.annotation.NonNull;
import android.support.v4.content.ContextCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;

import com.google.common.collect.Lists;

import net.syncthing.lite.R;
import net.syncthing.lite.databinding.ListviewDeviceBinding;

import it.anyplace.sync.core.beans.DeviceStats;

public class DevicesAdapter extends ArrayAdapter<DeviceStats> {

    public DevicesAdapter(Context context) {
        super(context, R.layout.listview_device, Lists.newArrayList());
    }

    @NonNull
    @Override
    public View getView(int position, View v, @NonNull ViewGroup parent) {
        ListviewDeviceBinding binding;
        if (v == null) {
            binding = DataBindingUtil.inflate(LayoutInflater.from(getContext()), R.layout.listview_device, parent, false);
        } else {
            binding = DataBindingUtil.bind(v);
        }
        DeviceStats deviceStats = getItem(position);
        binding.deviceName.setText(deviceStats.getName());
        int color = 0;
        switch (deviceStats.getStatus()) {
            case OFFLINE:
                color = R.color.device_offline;
                break;
            case ONLINE_INACTIVE:
                color = R.color.device_online_inactive;
                break;
            case ONLINE_ACTIVE:
                color = R.color.device_online_active;
                break;
        }
        // TODO: this is not working (and will also break on API 19
        binding.deviceIcon
                .setColorFilter(ContextCompat.getColor(getContext(), color), PorterDuff.Mode.SRC_IN);
        return binding.getRoot();
    }
}

File: app/src/main/kotlin/net/syncthing/lite/adapters/DevicesAdapter.kt
package net.syncthing.lite.adapters

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import net.syncthing.java.core.beans.DeviceInfo
import net.syncthing.lite.databinding.ListviewDeviceBinding
import kotlin.properties.Delegates

class DevicesAdapter: RecyclerView.Adapter<DeviceViewHolder>() {
    var data: List<DeviceInfo> by Delegates.observable(listOf()) {
        _, _, _ -> notifyDataSetChanged()
    }

    var listener: DeviceAdapterListener? = null

    init {
        setHasStableIds(true)
    }

    override fun getItemCount() = data.size
    override fun getItemId(position: Int) = data[position].deviceId.deviceId.hashCode().toLong()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = DeviceViewHolder(
            ListviewDeviceBinding.inflate(
                    LayoutInflater.from(parent.context), parent, false
            )
    )

    override fun onBindViewHolder(holder: DeviceViewHolder, position: Int) {
        val deviceStats = data[position]
        val binding = holder.binding

        binding.name = deviceStats.name
        binding.isConnected = deviceStats.isConnected

        binding.root.setOnLongClickListener { listener?.onDeviceLongClicked(deviceStats) ?: false }

        binding.executePendingBindings()
    }
}

interface DeviceAdapterListener {
    fun onDeviceLongClicked(deviceInfo: DeviceInfo): Boolean
}

class DeviceViewHolder(val binding: ListviewDeviceBinding): RecyclerView.ViewHolder(binding.root)